### DISCUSSION TITLE
CommandBar Padding

### DISCUSSION THREAD

**Discussion by gallwood** [BlazorFluentUI/BlazorFluentUI#391] [Q&A] [answered]
I see the CommandBar style has a left padding of 24px. Why is this? It seems to look odd when I use it at the top of my page :(

Is there a way I can override the left padding? I have tried surrounding the CommandBar in a <div> that has a style of padding-left: 0px !important; but this doesn't help.

BTW, I'm new to css ;)

Thanks

Graham

**Comment by limefrogyank** [comment, marked as answer]
Two options:
1.  If using css in a style tag, you'd need to be _more specific_ than the current style.  Currently, it's styled with a class name and an attribute selector.  I think if you try to do two selectors... like `div.ms-CommandBar`  (element type and class name) that might be enough to override the styling.  I'm not sure.
2. This is the better option.  All BlazorFluentUI elements should have the Style attribute mapped to the main element that starts the component.  So it should map to the style attribute on the .ms-Commanbar div.  Style attributes override css, so this is what you should use. (padding-left 0px;)

As for why, at the time, this is what the react version of this control used.  I'm not sure if that's still the case, but if it made sense to the UI designers for Fluent UI, that was good enough for me. :)

**Reply by gallwood** [reply]
Thanks for the options. 

Unfortunately, I still can't get either option to work.

Regarding option 2, setting the Style attribute isn't working I think because the code is applying the passing in Style to the outer component (ResizeGroup), not the actually inner div that uses the class ms-CommandBar. That's where my expertise ends :)

**Reply by limefrogyank** [reply]
Ok, the Style attribute won't work here... but you can still use plain css.  Just be more specific:
```
.ms-ResizeGroup div.ms-CommandBar {
    padding-left:0px;
}
```
This means:  looks at all elements with class "ms-ResizeGRoup", then drill down to a descendant that is a `div` element and has class 'ms-CommandBar.   It's more specific than what the isolated css has so it has priority.

**Reply by gallwood** [reply]
Thank you!

This worked perfectly. I have a couple of observations that I can't explain though. Maybe someone could.

1. This style has to be in the site.css file, it does not work when using css isolation (i.e. in a _component._.razor.css file). Not sure why.
2. Why is a further qualifier needed in the selector, the `.ms-ResizeGroup`? I have the CommandBar component inside more `div`'s but these don't seem to be needed in the selector.

I understand that the answer may be quite complex, so don't worry if it's too long to answer right now.

Thanks

**Reply by limefrogyank** [reply]
Css isolation only works on plain html on your razor page. It adds an ID attribute to all elements that also appears in the compiled css. It can't add that attribute to child components though, so none of the isolated css will work on child components.

However, you can use the css tag '::deep' to apply css to child components. The ID attribute will only apply to css to the left of the deep tag. Anything to the right of it will be plain css.

**Reply by limefrogyank** [reply]
I chose ms-ResizeGroup because all commandbars have them. Yes, it's unneeded normally. But, more specific rules have priority in styling. Having three things in the rule beats having two things in the css isolation file.

A space between two class names means any descendant of the first class. Immediate descendant would use '>' I think. No space between selectors means they are on the same element.

**Comment by gallwood** [comment]
Thank you. I’ll give it a go over the weekend.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
